File: ios/LocalNetworkPrivacy/LocalNetworkPrivacy.m
#import <UIKit/UIKit.h>
#import "LocalNetworkPrivacy.h"

@interface LocalNetworkPrivacy () <NSNetServiceDelegate>

@property (nonatomic) NSNetService *service;
@property (nonatomic) void (^completion)(BOOL);
@property (nonatomic) NSTimer *timer;
@property (nonatomic) BOOL publishing;

@end

@implementation LocalNetworkPrivacy

- (instancetype)init {
    if (self = [super init]) {
        self.service = [[NSNetService alloc] initWithDomain:@"local." type:@"_lnp._tcp." name:@"LocalNetworkPrivacy" port:1100];
    }
    return self;
}

- (void)dealloc {
    [self.service stop];
}

- (void)checkAccessState:(void (^)(BOOL))completion {
    self.completion = completion;

    self.publishing = YES;
    self.service.delegate = self;
    [self.service publish];

    self.timer = [NSTimer scheduledTimerWithTimeInterval:2 repeats:NO block:^(NSTimer * _Nonnull timer) {
        [self.timer invalidate];
        self.completion(NO);
    }];
}


#pragma mark - NSNetServiceDelegate

- (void)netServiceDidPublish:(NSNetService *)sender {
    [self.timer invalidate];
    self.completion(YES);
}

@end

File: ios/LocalNetworkPrivacy/LocalNetworkPrivacy.swift
//
//  LocalNetworkPrivacy.swift
//  CloudVoiceExpress
//
//  Created by Geraint White on 29/04/2022.
//

import Foundation
import UIKit

@objc public enum OptionalBool: Int {
    case none
    case yes
    case no
}

@objc open class LocalNetworkPrivacy : NSObject {
    let service: NetService

    var completion: ((Bool) -> Void)?
    var timer: Timer?
    var publishing = false

    static var granted: OptionalBool = OptionalBool.none

    @objc public override init() {
        service = .init(domain: "local.", type:"_lnp._tcp.", name: "LocalNetworkPrivacy", port: 1100)
        super.init()
    }

    @objc public static
    func authorizationStatus() -> OptionalBool {
        return LocalNetworkPrivacy.granted
    }

    @objc public
    func checkAccessState(completion: @escaping (Bool) -> Void) {
        self.completion = completion

        timer = .scheduledTimer(withTimeInterval: 2, repeats: true, block: { timer in
            guard UIApplication.shared.applicationState == .active else {
                return
            }

            if self.publishing {
                LocalNetworkPrivacy.granted = OptionalBool.no
                self.timer?.invalidate()
                self.completion?(false)
            }
            else {
                self.publishing = true
                self.service.delegate = self
                self.service.publish()

            }
        })
    }

    deinit {
        service.stop()
    }
}

extension LocalNetworkPrivacy : NetServiceDelegate {
    public func netServiceDidPublish(_ sender: NetService) {
        LocalNetworkPrivacy.granted = OptionalBool.yes
        timer?.invalidate()
        completion?(true)
    }
}
